src/jit/codegenlegacy.cpp

```diff
@@ -19036,16 +19036,31 @@ regMaskTP CodeGen::genCodeForCall(GenTreeCall* call, bool valUsed)
                     {
                         noway_assert(callType == CT_USER_FUNC);
 
-                        void* pAddr;
-                        addr = compiler->info.compCompHnd->getAddressOfPInvokeFixup(methHnd, (void**)&pAddr);
-                        if (addr != NULL)
+                        CORINFO_CONST_LOOKUP lookup;
+                        compiler->info.compCompHnd->getAddressOfPInvokeTarget(methHnd, &lookup);
+
+                        void* addr = lookup.addr;
+
+                        _ASSERTE(addr != NULL);
+
+                        bool direct;
+#if defined(_TARGET_ARM_)
+                        direct = compiler->codeGen->validImmForBL((ssize_t)addr);
+#else
+                        direct           = true;
+#endif
+
+                        if (lookup.accessType == IAT_VALUE || lookup.accessType == IAT_PVALUE)
                         {
 #if CPU_LOAD_STORE_ARCH
                             // Load the address into a register, indirect it and call  through a register
                             indCallReg = regSet.rsGrabReg(RBM_ALLINT); // Grab an available register to use for the CALL
                                                                        // indirection
                             instGen_Set_Reg_To_Imm(EA_HANDLE_CNS_RELOC, indCallReg, (ssize_t)addr);
-                            getEmitter()->emitIns_R_R_I(INS_ldr, EA_PTRSIZE, indCallReg, indCallReg, 0);
+
+                            if (lookup.accessType == IAT_PVALUE || (lookup.accessType == IAT_VALUE && !direct))
```

Comment:
`(lookup.accessType == IAT_VALUE && !direct)` is not right. If this condition was ever true, it would result into silent bad codegen, If you do not want to deal with it with in legacy backend  because of it is currently unreachable, add `noway_assert(lookup.accessType == IAT_VALUE && compiler->codeGen->validImmForBL((ssize_t)addr))` here instead. This will make sure that it won't generate bad code silently.